File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells, potions, success):
        pairs = []
        potions.sort()
        for spell in spells:
            count = 0
            searchVal = math.ceil(success / spell)
            if potions[-1] < searchVal:
                pairs.append(0)
                continue
            closeIdx = self.binarySearch(0, len(potions) - 1, searchVal, potions)
            count += len(potions) - closeIdx
            pairs.append(count)
        return pairs

    def binarySearch(self, lo, hi, search, array):
        if lo > hi:
            return hi + 1
        mid = ((hi - lo) // 2) + lo
        if array[mid] < search:
            return self.binarySearch(mid + 1, hi, search, array)
        else:
            return self.binarySearch(lo, mid - 1, search, array)

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.swift
class Solution {
    func successfulPairs(_ spells: [Int], _ potions: [Int], _ success: Int) -> [Int] {
        let sortedPotions = potions.sorted()
        let searchValues = spells.map { spell in
            return Int(ceil(Double(success) / Double(spell)))
        }
        var pairs = [Int](repeating: 0, count: spells.count)

        for (index, searchVal) in searchValues.enumerated() {
            let count = sortedPotions.count - binarySearch(sortedPotions, searchVal)
            pairs[index] = count
        }
        return pairs
    }

    func binarySearch(_ array: [Int], _ search: Int) -> Int {
        var low = 0
        var high = array.count - 1
        while low <= high {
            let mid = low + (high - low) / 2
            if array[mid] < search {
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        return low
    }
}
